### PR TITLE
Fix excluded IDs leaking into product search results

### PR DESCRIPTION
**Pull request by mordeth**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Product searches re-added excluded IDs after the SQL exclusion had already removed them: a matching variation contributed its parent through the `parent_id` column, and a numeric search term appended the searched ID and its parent unconditionally. Already selected variable products therefore came back in the Linked Products search, where clicking one silently removed it from the selection.

Excluded parents are now discarded in the SELECT, and the exclusion is applied inside the numeric-term branch, so nothing excluded reaches the result list. Adds regression coverage for parent, variation, and numeric-ID behavior.

Rebased onto #67420, which fixes the OR-group precedence in the same query. That fix is what lets the exclusion live entirely in SQL here.

Scope: the numeric-term branch now enforces `exclude`, but it still bypasses the `include` allowlist, the status filter, and the product type filter, as it always has. Those are pre-existing and left alone here.

Zero-valued `exclude` entries are filtered out during normalization. `absint()` maps any non-numeric value to `0`, which was inert against `posts.ID NOT IN(0)` but would have matched every top-level row through `posts.post_parent IN(0)`.

Backward compatibility: No public signature, hook, or filter changes. The existing `$exclude` argument is now honored consistently. Custom results returned by `woocommerce_product_pre_search_products` remain unchanged.

This does not change the `exclude` AJAX request format introduced by #67063.

Addresses #67293.

Bug introduced in PR #67063.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://developer.woocommerce.com/docs/contribution/testing/writing-high-quality-testing-instructions/), include your detailed testing instructions:

1. Edit a product and open the **Linked Products** section.
2. Select a variable product as an upsell or cross-sell.
3. Search for the selected variable product by name and confirm it does not appear in the results.
4. Search using the selected product's numeric ID and confirm it does not appear.
5. Remove the product from the selection and repeat the searches. Confirm the expected parent and variation results appear.
6. Run `pnpm --filter=@woocommerce/plugin-woocommerce test:php:env -- --filter WC_Product_Data_Store_CPT_Test`.

### Testing that has already taken place:

- PHPUnit in wp-env using PHP 8.1.34: 26 tests and 74 assertions passed. Both halves were mutation-checked: disabling the SELECT guard fails the variation test, disabling the numeric guards fails the numeric test.
- Browser verification confirmed that an excluded variable parent stays out of search results while its variation remains independently selectable.
- PHPStan completed without errors.
- Modified-file and full-branch PHP lint checks passed.
- An adversarial change-risk review found no public-contract, persistence, execution-context, or rollback concerns. It did surface that `WC_AJAX::json_search_downloadable_products_and_variations()` never passes its exclude list to `search_products()`, so the order screen's "Grant access" field is not covered. That predates this change and is deliberately left alone.
- Multisite was not tested. This change does not introduce site-scoped state or install-layout assumptions.

### Milestone

- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

Created manually.

</details>

### Release Communication

Select if this PR needs a generated summary for release notes:

- [ ] **Feature Highlight** - For user-facing features (what changed, user impact)
- [ ] **Developer Advisory** - For developer-facing changes (what changed, how to detect, actions needed)

<details>
<summary>When to use each?</summary>

**Feature Highlight**: New features, UI changes, or improvements that merchants/store owners will notice.
- Example: "New bulk editing for products", "Improved checkout performance"

**Developer Advisory**: Breaking changes, deprecations, or changes that affect themes/plugins/extensions.
- Example: "Hook signature change", "Deprecated filter", "REST API field removed"

An AI will analyze your PR and post a draft comment for you to review and edit.
</details>

### CODE REVIEW

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php [hunk 1]
```diff
@@ -2123,6 +2123,12 @@ public function search_products( $term, $type = '', $include_variations = false,
 			$product_ids[] = wp_get_post_parent_id( $post_id );
 		}
```
Comment (Aljullu) [inline plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php]: **For the numeric ID matches:**

Instead of removing the excluded product IDs below, could we check whether `$post_id` or the result of `wp_get_post_parent_id( $post_id )` match `exclude` inside this `if` block and then don't append them to `$product_ids`?

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php
Comment (Aljullu) [inline plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php]: **For the parent ID matches:**

If we could go with a SQL-only solution, I believe it would be the cleanest and most performance way to solve this.

I might be missing some edge cases, but would something like this work?

```PHP
		if ( $exclude_ids ) {
			$exclude_id_list = implode( ',', $exclude_ids );
			$parent_id_select = "CASE WHEN posts.post_parent IN ({$exclude_id_list}) THEN NULL ELSE posts.post_parent END as parent_id";
		}
```

```suggestion
			"SELECT DISTINCT posts.ID as product_id, {$parent_id_select} as parent_id FROM {$wpdb->posts} posts
```

**Review by Aljullu — COMMENTED** [review]
Thanks for following up on this, @mordeth! I confirm this is working as expected, I completely missed those cases when testing https://github.com/woocommerce/woocommerce/pull/67063!

I left a couple of suggestions for alternatives to solve this. I think they are a bit better because they filter the results earlier (SQL level, or PHP before addition to the array). But looking forward to hear what you think. I might also have missed some edge cases.

**Comment by mordeth** [comment]
Thanks for confirming the cases, Albert.

Both land on the same results as the current guard, and the tests pass either way. So it's about shape.

I'd keep the single guard before the `return`: one rule, one place, at the one exit. Split in two, a future append path can miss it. And the set is `$limit`-capped at 30, so there's no perf win to chase.

Happy to modify it if you are seeing something.

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php
Comment (mordeth) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php]: Mechanics are fine: `NULL` becomes `0` via `wp_parse_id_list()`, and `0` is already there from every top-level `post_parent`. It would need the `absint` mapping the `NOT IN` above uses, though.

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 1 above)
Comment (mordeth) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php]: Works. My only hesitation is that it splits the rule in two, same as the other thread.

**Review by Aljullu — COMMENTED** [review]
Thanks for checking, good point! We can look for a second opinion if needed, I understand that my suggestion means we need to handle the excluded items in every request we do to the database, which might be a bit more complex than simply querying all products and filtering them at the end. But I think that's cleaner because we know excluded items are never added to the `$product_ids` array. It's also aligned with the current way of handling the `exclude` param:

https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php#L2077-L2079

> And the set is `$limit`-capped at 30, so there's no perf win to chase.

Now that you mention that, would the `$limit` be applied correctly if we are filtering out the items _after_ the SQL query? I wouldn't say it's a blocker, but IMO it's another signal that going with the SQL solution might be cleaner.

Sidenote: in fact, apparently we also have a bug that numeric matches are appended to the `$product_ids` even if the array length is already above the limit. But I think that's a very specific edge case, so not a big deal. :smile:

**Comment by mordeth** [comment]
Good question on the limit. I checked and it holds either way: `LIMIT` applies to rows, and excluded products are already dropped by the `NOT IN`, so they never take a slot. The only thing filtered afterwards is the parent IDs, which come from a derived column and were never counted against the limit (the function already returns up to 2x `$limit`). So the counts come out the same.

That said, the consistency argument convinces me. Moved both filters up in 8f112f16e9: `CASE` in the SELECT with the `absint` mapping, and the check inside the numeric block. Tests cover it either way, and I mutation-checked both halves to confirm they still fail when each guard is disabled.

Good spot on the numeric appends bypassing the limit. Separate from this fix, so I'll leave it be unless you'd rather I fold it in.

**Comment by mordeth** [comment]
Thanks for the feedback, Albert.

First a correction on my earlier comment: `NULL` does not become `0`. `wp_parse_list()` filters non-scalars with `array_filter( $list, 'is_scalar' )`, so it is dropped. Same result, but my explanation was wrong.

I also found something while testing. If the term contains " or ", `$search_where` is built as `AND (group1) OR (group2)` without wrapping parentheses. Everything appended after it, including `NOT IN`, applies to the last group only. So a product matched by the first group is not excluded.

Searching "Alpha or Beta" with Alpha excluded:

- trunk: Alpha, Beta
- SQL only: Alpha, Beta
- post-query guard: Beta

This is not a regression, it is there since 3.6.0 when `exclude` was added, and it affects `include`, status and type as well. But with the SQL only version this PR would still claim more than it fixes.

So I reverted to the guard and added a test for both OR groups. It fails with the SQL only version. I will open a separate issue for the parentheses, since fixing that also changes status and type results for OR searches.

**Review by Aljullu — COMMENTED** [review]
Good finding with that issue!

In that case, I think we should fix https://github.com/woocommerce/woocommerce/issues/67392 first. IMO the fact that the current codebase has that issue shouldn't influence how we fix this separate issue. What do you think?

**Comment by mordeth** [comment]
Agreed, that is the right order.

I opened #67420 for it. It wraps the disjunction in its own parentheses so include, exclude, status and type apply to every group, with regression coverage for each leak.

Once that is merged I will rebase this on top and move the filtering into SQL as you suggested, since the OR path stops leaking then.

Happy to park this one until #67392 is closed.

**Comment by mordeth** [comment]
Thanks for the feedback, @Aljullu.

#67420 is merged, so I rebased on top and moved the filtering into SQL as you suggested. With the parentheses fix in place the OR path no longer leaks, so the guard is gone.

One extra: `absint()` maps any non-numeric exclude value to `0`, the `post_parent` of every top-level product, so `exclude[]=abc` matched every row. Zeros are filtered out now.

Verified red to green. Branch was rebuilt, so the history is new.

File: plugins/woocommerce/tests/php/includes/data-stores/class-wc-product-data-store-cpt-test.php [hunk 2]
```diff
@@ -225,6 +225,107 @@ public function test_search_products_or_term_returns_all_groups(): void {
 		$this->assertContains( $beta->get_id(), $results, 'Product matched by the second OR group should be returned' );
 	}
 
+	/**
+	 * @testdox Excluded variable products should not be re-added to search results by their matching variations.
+	 */
+	public function test_search_products_excludes_variable_products_with_matching_variations() {
+		$parent = new WC_Product_Variable();
+		$parent->set_name( 'Excludable variable widget' );
```
Comment (Aljullu) [inline plugins/woocommerce/tests/php/includes/data-stores/class-wc-product-data-store-cpt-test.php]: Several tests use `widget` in the product name. As mentioned in the other PR, not a blocker from my side. But I don't really like it. 😄 

```suggestion
		$parent->set_name( 'Excludable variable product' );
```

File: plugins/woocommerce/tests/php/includes/data-stores/class-wc-product-data-store-cpt-test.php [hunk 3]
```diff
@@ -225,6 +225,107 @@ public function test_search_products_or_term_returns_all_groups(): void {
 		$this->assertContains( $beta->get_id(), $results, 'Product matched by the second OR group should be returned' );
 	}
 
+	/**
+	 * @testdox Excluded variable products should not be re-added to search results by their matching variations.
+	 */
+	public function test_search_products_excludes_variable_products_with_matching_variations() {
```
Comment (Aljullu) [inline plugins/woocommerce/tests/php/includes/data-stores/class-wc-product-data-store-cpt-test.php]: Not a big deal, but in the tests added in the other PR we add return types. Should we keep it consistent?


```suggestion
	public function test_search_products_excludes_variable_products_with_matching_variations(): void {
```

**Review by Aljullu — APPROVED** [review]
Thanks for the updates, @mordeth! This is testing as described and code looks good. I left a couple of minor nits in tests, but LGTM.
